src/esphomelib/output/my9231_output_component.cpp

```diff
@@ -0,0 +1,189 @@
+  ESP_LOGCONFIG(TAG, "    Command: 0x%02X", command);
+
+  this->init_chips(command);
+  ESP_LOGV(TAG, "Chips initialized.");
+  this->loop();
+}
+
+void MY9231OutputComponent::loop() {
+  if (!this->update_) {
+    return;
+  }
+
+  for (auto pwm_amount : this->pwm_amounts_) {
+    this->write_word(pwm_amount, this->bit_depth_);
+  }
+  // Send 8 DI pulses. After 8 falling edges, the duty data are store.
+  this->send_di_pulses(8);
+  this->update_ = false;
+}
+
+MY9231OutputComponent::Channel *MY9231OutputComponent::create_channel(
+    uint8_t channel, PowerSupplyComponent *power_supply, float max_power) {
+  ESP_LOGV(TAG, "Getting channel %d...", channel);
+  auto *c = new Channel(this, channel);
+  c->set_power_supply(power_supply);
+  c->set_max_power(max_power);
+  return c;
+}
+
+float MY9231OutputComponent::get_setup_priority() const {
+  return setup_priority::HARDWARE;
+}
+
+void MY9231OutputComponent::set_channel_value(uint8_t channel, uint16_t value) {
+  ESP_LOGV(TAG, "set channels %u to %u", channel, value);
+  if (this->pwm_amounts_[channel] != value) {
+    this->update_ = true;
+  }
+  this->pwm_amounts_[this->num_channels_ - channel - 1] = value;
+}
+
+void MY9231OutputComponent::set_num_channels(uint16_t num_channels) {
+  this->num_channels_ = num_channels;
+}
+
+uint16_t MY9231OutputComponent::get_num_channels() const {
+  return this->num_channels_;
+}
+
+void MY9231OutputComponent::set_num_chips(uint8_t num_chips) {
+  this->num_chips_ = num_chips;
+}
+
+uint8_t MY9231OutputComponent::get_num_chips() const {
+  return this->num_chips_;
+}
+
+void MY9231OutputComponent::set_bit_depth(uint8_t bit_depth) {
+  this->bit_depth_ = bit_depth;
+}
+
+uint8_t MY9231OutputComponent::get_bit_depth() const {
+  return this->bit_depth_;
+}
+
+void MY9231OutputComponent::set_update(bool update) { this->update_ = update; }
+
+uint16_t MY9231OutputComponent::get_max_amount() const {
+  return (uint32_t(1) << this->bit_depth_) - 1;
+}
+
+void MY9231OutputComponent::init_chips(uint8_t command) {
+  // Send 12 DI pulse. After 6 falling edges, the duty data are stored
+  // and after 12 rising edges the command mode is activated.
+  this->send_di_pulses(12);
+  delayMicroseconds(12);
+  for (uint8_t i = 0; i < this->num_chips_; i++) {
+    this->write_word(command, 8);
+  }
+  // Send 16 DI pulse. After 14 falling edges, the command data are
+  // stored and after 16 falling edges the duty mode is activated.
+  this->send_di_pulses(16);
+}
+
+void MY9231OutputComponent::write_word(uint16_t value, uint8_t bits) {
+  for (uint8_t i = bits; i > 0; i--) {
+    this->pin_di_.digital_write(value & (1 << (i - 1)));
+    this->pin_dcki_.digital_write(!this->pin_dcki_.digital_read());
+  }
+}
+
+void MY9231OutputComponent::send_di_pulses(uint8_t count) {
+  delayMicroseconds(12);
+  for (uint8_t i = 0; i < count; i++) {
+    this->pin_di_.digital_write(true);
+    this->pin_di_.digital_write(false);
```

Comment:
What is the timing requirement of these pulses? In theory (if all optimizations are applied), this loop could run at around 20-40MHz.

Reply:
Ok, let's just leave it as is then 👍